### PR TITLE
feat(eslint-plugin): [no-misused-promises] add granular options within `checksVoidReturns`

### LINKED ISSUE #4619: [no-misused-promises] async react callback properties when void expected
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "no-misused-promises": ["error"]
  }
}
```

```TSX
			<button
				type="button"
				onClick={async () => {
					await doSomethingAsync();
				}}
			>
				action
			</button>
```

or

```tsx

			<button type="button" onClick={doSomethingAsync}>
				action
			</button>
```

**Expected Result**

No linting error.

**Actual Result**

A linting error is being reported:

```
src/app/App.tsx
  129:13  error  Promise-returning function provided to attribute where a void return was expected  @typescript-eslint/no-misused-promises
```

One way to avoid this error would be to wrap the callback's content in an async IIFE:

```tsx
			<button
				type="button"
				onClick={() => {
					(async () => {
						await doSomethingAsync();
					})();
				}}
			>
				action
			</button>
```

But that's not really practical when the callback is an existing function (like the 2nd example above).

**Additional Info**

I guess this new behavior is kind of expected, but it's probably gonna breaks linting on a lot of React projects (including ours, which has now over 100 erros following this update). I'll welcome any suggestion as to how to address it without simply disabling the rule (and risk missing some potential bugs that were detected before).

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.13.0` |
| `@typescript-eslint/parser`        | `5.13.0` |
| `TypeScript`                       | `4.5.5` |
| `ESLint`                           | `8.10.0` |
| `node`                             | `16.13.2` |

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
## PR Checklist

-   [x] Addresses an existing issue: fixes #4619
-   [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
-   [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

Allows user configurations to provide an object containing opt-outs for:

```ts
interface ChecksVoidReturnOptions {
  arguments?: boolean;
  attributes?: boolean;
  properties?: boolean;
  returns?: boolean;
  variables?: boolean;
}
```

Specifically the `arguments` one should be useful for folks #4619. Folks who don't mind passing a `() => Promise<void>` to a `() => void` handler, such as React button clicks, can disable it with:

```json
{
  "@typescript-eslint/no-misused-promises": [
    "error",
    {
      "checksVoidReturn": {
        "arguments": false
      }
    }
  ]
}
```

...though we should note per https://github.com/typescript-eslint/typescript-eslint/issues/4619#issuecomment-1055614155 that doing so is silencing a legitimate report about a misused promise. The "proper" (safe) behavior would be to create some kind of `wrapAsyncFunction` as mentioned in https://github.com/typescript-eslint/typescript-eslint/issues/4619#issuecomment-1055653098.

### CODE REVIEW

**Comment by JoshuaKGoldberg** [comment]
👋 @uhyo, just a heads up following #4541

File: packages/eslint-plugin/docs/rules/no-misused-promises.md [hunk 1]
```diff
@@ -45,15 +67,17 @@ expected to be checked, your configuration will look like this:
 }
 ```
 
-Likewise, if you don't want to check conditionals, you can configure the rule
-like this:
+You can disable selective parts of the `checksVoidReturn` option by providing an
+object that disables specific checks:
```
Comment (bradzacher) [inline packages/eslint-plugin/docs/rules/no-misused-promises.md:71]: we should probably add a quick description / example for each option.
it's verbose but will help people self-service

**Review by bradzacher — APPROVED** [review]
implementation looks good - just one comment around documentation

**Review by bradzacher — APPROVED** [review]
NOICE

**Pull request merged**